crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -460,17 +463,184 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+    let (upgraded_package_modules, new_package_id) =
+        publish_and_verify_modules::<_, _, Mode>(context, module_bytes)?;
+
+    // Check that this package ID points to a package and get the package we're upgrading.
+    let current_package = fetch_package(context, &upgrade_ticket.package.bytes)?;
+
+    // Full backwards compatibility except that we allow friend function signatures to change.
+    let compatiblity_checker = {
+        let check_struct_and_pub_function_linking = true;
+        let check_struct_layout = true;
+        let check_friend_linking = false;
+        Compatibility::new(
+            check_struct_and_pub_function_linking,
+            check_struct_layout,
+            check_friend_linking,
+        )
+    };
+    check_compatibility(
+        &current_package,
+        &upgraded_package_modules,
+        compatiblity_checker,
+        upgrade_ticket.policy,
+    )?;
+
+    // Deduplicate any dependency ids.
+    dep_ids.dedup();
+
+    // Read the package dependencies.
+    let dependency_packages = dep_ids
+        .iter()
+        .map(|id| fetch_package(context, id))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let upgraded_object_id = context.upgrade_package(
+        &current_package,
+        new_package_id,
+        upgraded_package_modules,
+        &dependency_packages,
+    )?;
+
+    Ok(vec![Value::Object(ObjectValue::new(
+        UpgradeReceipt::type_().into(),
+        /* has_public_transfer */ true,
+        /* used_in_non_entry_move_call */ false,
+        &bcs::to_bytes(&UpgradeReceipt::new(upgrade_ticket, upgraded_object_id)).unwrap(),
+    )?)])
+}
+
+// Load the package and do any error handling that is needed.
+// We load them one at a time so we can determine which package id we were unable to load.
+fn fetch_package<'a, E: fmt::Debug, S: StorageView<E>>(
+    context: &'a ExecutionContext<E, S>,
+    package_id: &ObjectID,
+) -> Result<MovePackage, ExecutionError> {
+    match context.state_view.get_package(package_id) {
+        Err(e) => Err(ExecutionError::new_with_source(
+            ExecutionErrorKind::PackageUpgradeError {
+                upgrade_error: PackageUpgradeError::UnableToFetchPackage {
+                    package_id: *package_id,
+                },
+            },
+            e,
+        )),
+        Ok(None) => Err(ExecutionError::from_kind(
+            ExecutionErrorKind::PackageUpgradeError {
+                upgrade_error: PackageUpgradeError::UnableToFetchPackage {
+                    package_id: *package_id,
+                },
+            },
+        )),
+        Ok(Some(pkg)) => Ok(pkg),
+    }
+}
+
+fn check_compatibility<'a>(
+    existing_package: &MovePackage,
+    upgrading_modules: impl IntoIterator<Item = &'a CompiledModule>,
+    compatiblity_checker: Compatibility,
+    policy: u8,
+) -> Result<(), ExecutionError> {
+    // Make sure this is a known upgrade policy.
+    if !is_valid_package_upgrade_policy(&policy) {
```

Comment:
nit maybe make `is_valid_package_upgrade_policy` return an enum?

Reply:
(note, as I've recently been down this path) I originally tried to expose this as a Rust enum, without any constants, but this unfortunately does not work because even if your Rust enum definition specifies specific numbers for enumerands, BCS will serialize them as `0`, `1`, `2`, etc.